src/connectionbenchmark/connectionbenchmark.cpp

```diff
@@ -0,0 +1,136 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "benchmarktaskdownload.h"
+#include "benchmarktaskping.h"
+#include "connectionbenchmark.h"
+#include "connectionhealth.h"
+#include "controller.h"
+#include "constants.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "mozillavpn.h"
+#include "taskscheduler.h"
+
+namespace {
+Logger logger(LOG_MODEL, "ConnectionBenchmark");
+}
+
+ConnectionBenchmark::ConnectionBenchmark() {
+  MVPN_COUNT_CTOR(ConnectionBenchmark);
+}
+
+ConnectionBenchmark::~ConnectionBenchmark() {
+  MVPN_COUNT_DTOR(ConnectionBenchmark);
+}
+
+void ConnectionBenchmark::setConnectionSpeed() {
+  logger.debug() << "Set speed";
+
+  if (m_download >= Constants::BENCHMARK_THRESHOLD_SPEED_FAST) {
+    m_speed = SpeedFast;
+  } else if (m_download >= Constants::BENCHMARK_THRESHOLD_SPEED_MEDIUM &&
+             m_download < Constants::BENCHMARK_THRESHOLD_SPEED_FAST) {
+    m_speed = SpeedMedium;
+  } else {
+    m_speed = SpeedSlow;
+  }
+
+  emit speedChanged();
+  setState(StateReady);
+
+  m_benchmarkTasks.clear();
+}
+
+void ConnectionBenchmark::setState(State state) {
+  logger.debug() << "Set state" << state;
+  m_state = state;
+
+  emit stateChanged();
+}
+
+void ConnectionBenchmark::start() {
+  logger.debug() << "Start connection benchmarking";
+
+  MozillaVPN* vpn = MozillaVPN::instance();
+  Q_ASSERT(vpn);
+
+  Controller* controller = vpn->controller();
+  Controller::State controllerState = controller->state();
+  Q_ASSERT(controllerState == Controller::StateOn);
+
+  m_connectionHealth = vpn->connectionHealth();
+  if (m_connectionHealth->stability() == ConnectionHealth::NoSignal) {
+    handleStabilityChange();
+  }
+
+  connect(controller, &Controller::stateChanged, this,
+          &ConnectionBenchmark::stop);
+  connect(m_connectionHealth, &ConnectionHealth::stabilityChanged, this,
+          &ConnectionBenchmark::handleStabilityChange);
+
+  // Create ping benchmark
+  m_pingBenchmarkTask = new BenchmarkTaskPing();
+  connect(m_pingBenchmarkTask, &BenchmarkTaskPing::finished, this,
+          &ConnectionBenchmark::pingBenchmarked);
+  m_benchmarkTasks.append(m_pingBenchmarkTask);
+  TaskScheduler::scheduleTask(m_pingBenchmarkTask);
+
```

Comment:
You need to remove the task from `m_benchmarkTasks` when it's completed.
I would do this:

`connect(pingTask, &Task::completed, this, [this, pingTask]() { m_benchmarkTasks.remove(pingTask); });`